File: examples/calibration/evolution/sources/harvest.pl
package Marker;

sub Marker::new {
    my ($class, %options) = @_;
    return bless {
        current => undef,
        history => [],
        cursor => 0,
        fail_history => $options{fail_history},
        fail_cursor => $options{fail_cursor},
    }, $class;
}

sub Marker::write_current {
    my ($self, $key, $value) = @_;
    $self->{current} = { key => $key, value => $value };
}

sub Marker::write_history {
    my ($self, $key) = @_;
    die "history" if $self->{fail_history};
    push @{$self->{history}}, $key;
}

sub Marker::advance_cursor {
    my ($self) = @_;
    die "cursor" if $self->{fail_cursor};
    $self->{cursor}++;
}

sub Marker::cursor {
    my ($self) = @_;
    return $self->{cursor};
}

sub Marker::apply {
    my ($state, $key, $value) = @_;
    $state->write_current($key, $value);
    $state->write_history($key);
    $state->advance_cursor();
    return $state->cursor();
}

1;

File: examples/calibration/evolution/sources/harvest.py
class Ledger:
    def __init__(self, fail_index=False, fail_history=False):
        self.current = None
        self.index = {}
        self.history = []
        self.cursor = 0
        self.fail_index = fail_index
        self.fail_history = fail_history

    def write_current(self, key, value):
        self.current = (key, value)

    def write_index(self, key):
        if self.fail_index:
            raise OSError("index")
        self.index[key] = len(self.history)

    def write_history(self, key):
        if self.fail_history:
            raise OSError("history")
        self.history.append(key)

    def advance_cursor(self):
        self.cursor += 1


def advance_record(state, key, value):
    state.write_current(key, value)
    state.write_index(key)
    state.write_history(key)
    state.advance_cursor()
    return state.cursor
